site/docs/plugins/menu.md

```diff
@@ -93,57 +93,373 @@ bot.start();
+It will then be compared to see if the menu is outdated before any handler is run.
+
+It is thinkable that your menu changes, but all of the above things are staying the same.
+This is usually not the case, but if you are creating a menu that has this property, you should use a fingerprint function.
+
+```ts
+function ident(ctx: Context): string {
+  // return string that would change when your menu changes
+}
+const menu = new Menu("id", { fingerprint: (ctx) => ident(ctx) });
+```
+
+The fingerprint string will be added to the above list of things that impact the hash generation.
+That way, you can make sure that outdated menus are always detected.
+
+## How Does It Work
+
+The menu plugin works completely without storing any data.
+This is important for large bots with millions of users.
+Saving the states of all menus would consume too much memory.
+
+When you create your menu objects and link them together via `register` calls, no menus are actually built.
+Instead, the menu plugin will remember how to assemble new menus based on your operations.
+Whenever a menus is sent, it will replay these operation to render your menu.
+This includes layouting all dynamic ranges and generating all dynamic labels.
+Once the menu is sent, the rendered button array will be forgotten again.
+
+When a user presses a button of the menu, we need to find the handler that was added to that button at the time the menu was rendered.
+Hence, we simply render the old menu again.
+However, this time, we don't actually need the full layout—all we need is the overall structure, and that one specific button.
+Consequently, the menu plugin will perform a shallow rendering in order to be more efficient.
+In other words, the menu will only be rendered partially.
+
+Once the pressed button is known again (and we have checked that the menu is not [outdated](#outdated-menus-and-fingerprints)), we invoke the handler.
+
+Internally, the menu plugin makes heavy use of [API Transformer Functions](/advanced/transformers.md), e.g. to quickly render outgoing menus on the fly.
 
-TODO: explain the problem and the solution, mention shortcomings, explain config
+When you register the menus in a large hierachy for navigating, they will in fact not store these references explicitly.
+Under the hood, all menus of that one structure are added to the same large pool, and that pool is shared across all contained instances.
+Every menu is responsible for every other one in the index, and they can handle and render each other.
+(Most often, it is only the root menu that is actually passed to `bot.use` and receiving any updates.
+In that case, this one instance will handle the complete pool.)
+As a result, you are able to navigate between arbitrary menus without limits, and still the update handling can happen in `O(1)` because there is no need to search through hierarchies after the right menu to handle a button click.
```

Comment:
Not sure how ` are handled in hyperlinks, will this work?
```suggestion
As a result, you are able to navigate between arbitrary menus without limit, all while the update handling can happen in [`O(1)` time complexity](https://en.wikipedia.org/wiki/Time_complexity#Constant_time) because there is no need to search through entire hierarchies to find the right menu to handle any given button click.
```

Reply:
Yes.